rena.py

```diff
@@ -194,18 +213,134 @@ def recursive_nearest_agglomeration(data_matrix, connectivity, n_clusters,
+    """
+    ReNA is useful.
+
+    Parameters
+    ----------
+    masker: dd
 
+    n_cluster: int, optional (default 2)
+        Number of clusters.
 
-def approximate_data(X, labels):
-    _, inverse = np.unique(labels, return_inverse=True)
-    return X[..., inverse]
+    connectivity
 
+    scaling: bool, optional (default False)
+
+    memory: instance of joblib.Memory or string
+        Used to cache the masking process.
+        By default, no caching is done. If a string is given, it is the
+        path to the caching directory.
+
+    n_iter: int, optional (default 10)
+        Number of iterations of the recursive nearest agglomeration
+
+    n_jobs: int, optional (default 1)
+        Number of jobs in solving the sub-problems.
+
+    thr: float in the opened interval (0., 1.), optional (default 1e-7)
+        Threshold used to deal with eccentricities.
+
+    Attributes
+    ----------
+    `labels_`: numpy array
+
+    `n_clusters_`: int
+        Number of clusters
+
+    `sizes_`: numpy array
+        It contains the size of each cluster
+
+    """
+    def __init__(self, n_clusters=2, connectivity=None, masker=None, memory=None,
+                 scaling=False, n_iter=10, thr=1e-7, n_jobs=1):
+        self.n_clusters = n_clusters
+        self.connectivity = connectivity
+        self.memory = memory
+        self.scaling = scaling
+        self.n_iter = n_iter
+        self.n_jobs = n_jobs
+        self.masker = masker
+        self.thr = thr
+
+    def fit(self, X):
+        """Compute clustering of the data
+
+        Parameters
+        ----------
+        X : 2D array
+        """
+
+        X = check_array(X, ensure_min_features=2)
+
+        memory = self.memory
+        if isinstance(memory, six.string_types):
+            memory = Memory(cachedir=memory, verbose=0)
+
+        if self.n_clusters <= 0:
+            raise ValueError("n_clusters should be an integer greater than 0."
+                             " %s was provided." % str(self.n_clusters))
+
+        n_labels, labels = recursive_nearest_agglomeration(
+            self.masker, X, self.n_clusters, n_iter=self.n_iter, thr=self.thr)
+
+        sizes = np.bincount(labels)
+        sizes = sizes[sizes > 0]
+
+        self.labels_ = labels
+        self.n_clusters_ = np.unique(self.labels_).shape[0]
+        self.sizes_ = sizes
+        self.n_features = X.shape[1]
+        
+        return self
+
+
+    def transform(self, X):
+        """Apply clustering, reduce the dimensionality of the data
+
+        Parameters
+        ----------
+        X: 2D array
+        """
+        N = X.shape[0]
+        check_is_fitted(self, 'labels_')
+        #unique_labels = np.unique(self.labels_)
+
+#         nX = []
+#         for l in unique_labels:
+#             nX.append(np.mean(X[:, self.labels_ == l], axis=1))
+#         Xred =  np.array(nX).T
+        Xred = np.array([np.bincount(self.labels_, X[i,:])/np.bincount(self.labels_) for i in range(N)])
+
```

Comment:
do you need to calculate np.bincount(self.labels_) again?, It seems that they are already saved in self.sizes_

Reply:
Oh you are right. Fixed it.